### DISCUSSION TITLE
pybindings for nalgebra or other matrix math libs

### DISCUSSION THREAD

**Discussion by sjbeskur** [PyO3/pyo3#2029] [Questions] [answered]
I'm able to expose simple functions written in Rust to python using pyo3 but can see no way to expose complex "eigen" / matrix types. Does anyone know if this is possible?   I was hoping it would work like the pybind11 stuff in C++ that includes eigen vector / matrix types.

lib.rs

```
extern crate nalgebra as na;    
use pyo3::prelude::*;
use na::{SMatrix, Vector3, Matrix3};

/// Formats the sum of two numbers as string.
#[pyfunction]
fn sum_as_string(a: usize, b: usize) -> PyResult<String> {
    Ok((a + b).to_string())
}


#[pyfunction]
fn matrix_math(v3: Vector3<f64>, m3x3: Matrix3<f64>) -> PyResult<SMatrix<f64, 3, 1>>{
    let mxv = m3x3 * v3;
    Ok(mxv)
}

// Neiter way works

type Matrix3f = Matrix3<f64>;
type Vector3f = Vector3<f64>;

#[pyfunction]
fn matrix_math_w_types(v3: Vector3f, m3x3: Matrix3f) -> PyResult<SMatrix<f64, 3, 1>>{
    let mxv = m3x3 * v3;
    mxv
}

#[pymodule]
fn libpytest(_py: Python, m: &PyModule) -> PyResult<()> {
    m.add_function(wrap_pyfunction!(sum_as_string, m)?)?;
    m.add_function(wrap_pyfunction!(matrix_math, m)?)?;

    Ok(())
}
```

**Comment by adamreichold** [comment, marked as answer]
I think [`rust-numpy`](https://github.com/PyO3/rust-numpy) which provides glue between Python's [NumPy](https://github.com/numpy/numpy/) and Rust's [`ndarray`](https://github.com/rust-ndarray/ndarray) might be helpful in this case?

**Reply by davidhewitt** [reply]
To add to this, if you want to use `nalgebra` instead of `ndarray` a pull request to `rust-numpy` adding an optional feature is welcome!

**Reply by sjbeskur** [reply]
@adamreichold , @davidhewitt Thanks. I had not explored ndarray but will look into this.  nalgebra just had such great documentation I thought it might be the rust matrix math defacto.  Will definitely look into ndarray and rust-numpy.
